### PR TITLE
feat: ftl-fundamentals main and bonus goals

### PR DESCRIPTION
**Pull request by maiku1008**

### CODE REVIEW

File: calculator.go [hunk 1]
```diff
@@ -1,13 +1,125 @@
...
-return a + b
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+)
+
+// Add takes a list of numbers and returns the result of adding them together.
+func Add(nums ...float64) float64 {
```
Comment (bitfield) [inline calculator.go]: This is great, but one potential problem is that we could pass zero or one arguments to Add, which doesn't really makes sense! Could we write the parameter list in such a way that the compiler will require _at least two_ arguments, and maybe more?

File: calculator.go [hunk 2]
```diff
@@ -1,13 +1,125 @@
...
+	switch len(nums) {
+	case 0:
+		return 0
```
Comment (bitfield) [inline calculator.go]: This special case isn't really needed, is it? If there are no `nums` then the `for` loop will execute zero times, which is what we want! It's not that this code is wrong—it isn't. It's just that if we remove it, the tests will still pass, so that's a good way to define 'unnecessary code', isn't it?

File: calculator.go [hunk 3]
```diff
@@ -1,13 +1,125 @@
...
+	case 1:
```
Comment (bitfield) [inline calculator.go]: And the same applies here!

File: calculator.go [hunk 4]
```diff
@@ -1,13 +1,125 @@
...
+		return nums[0]
+	default:
```
Comment (bitfield) [inline calculator.go]: So if we remove the special cases, we don't even need the `switch` statement, do we?

File: calculator.go [hunk 5]
```diff
@@ -1,13 +1,125 @@
...
+// Divide takes a list of numbers and divides them.
+// Error is returned when dividing by 0.
+func Divide(nums ...float64) (float64, error) {
+	switch len(nums) {
+	case 0:
+		return 0, nil
+	case 1:
+		return nums[0], nil
+	default:
+		var product float64 = nums[0]
```
Comment (bitfield) [inline calculator.go]: `product` doesn't seem like quite the right name, does it? How about something more general like `result`?

File: calculator.go [hunk 6]
```diff
@@ -1,13 +1,125 @@
...
+		}
+		return product, nil
+	}
+}
+
+// Sqrt finds the square root of a number
+// Running this on 0 or negative numbers returns an error
+func Sqrt(num float64) (float64, error) {
+	if num < 0 {
+		return 0, fmt.Errorf("not allowed to find the square root of a negative numbers")
```
Comment (bitfield) [inline calculator.go]: And it would be polite to include the problematic number, wouldn't it?

File: calculator.go [hunk 7]
```diff
@@ -1,13 +1,125 @@
...
+}
+
+// StringMath takes a string, parses it and runs the math operation found therein
+// Returns an error if it cannot do so
+func StringMath(input string) (float64, error) {
+	var a, b float64
+	var operation string
+	var err error
+
+	operations := []string{"+", "-", "*", "/"}
```
Comment (bitfield) [inline calculator.go]: We might try using `fmt.Sscanf` or similar here, to do this a bit more compactly.

File: calculator_test.go [hunk 8]
```diff
@@ -2,23 +2,189 @@ package calculator_test
...
+		{"one zero", []float64{0, 3}, 0, true},
+		{"two zeroes", []float64{0, 0}, 0, true},
+		{"decimal", []float64{2.5, 4}, 0.625, false},
+		{"empty slice", []float64{}, 0, false},
+		{"one item in slice", []float64{3}, 3, false},
+		{"many numbers", []float64{1024, 2, 2, 2, 2, 2, 2, 2, 2, 2}, 2, false},
+	}
+	for _, tc := range testCases {
+		got, err := calculator.Divide(tc.nums...)
+		if tc.errExpected {
```
Comment (bitfield) [inline calculator_test.go]: And we can refactor this code to be more concise and handle _both_ unexpected error status cases, can't we? Can you see how? (Hint: suppose we had some `bool` variable `errReceived`, telling us whether or not we received an error from `Divide`. That would make the `if` statement easy to write, wouldn't it? How could we set such a variable?)

**Review by bitfield — COMMENTED** [review]
Great progress so far!

File: calculator_test.go (hunk 8 above)
Comment (micuffaro) [reply calculator_test.go:112]: Thanks for the feedback.

I'm not sure I completely understand what you're asking here.

Do we want to assert specific errors here, coming from upstream functions (in this case, `Divide()`)? 

In that case, one way that comes to mind is to make `StringMath()` return the extra `bool` value that tells us whether errors are from itself or from other functions it is calling.

It feels like extra machinery though...

**Comment by maiku1008** [comment]
Thanks John, I addressed most comments aside from the last one.

File: calculator_test.go (hunk 8 above)
Comment (maiku1008) [reply calculator_test.go]: I tried to implement this in [this commit](https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/e41009eb2149320aaa62c3a7b21dd902d968c63b), although frankly I don't see how my approach is adding anything useful to the current `StringMath()` test, as I can probably leave it at asserting the `!tc.errExpected` block.

File: calculator.go [hunk 9]
```diff
@@ -1,13 +1,96 @@
...
-return a + b
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+// Add takes a list of numbers and returns the result of adding them together.
+func Add(a, b float64, nums ...float64) float64 {
+	var result float64 = a + b
```
Comment (bitfield) [inline calculator.go]: Great! And we could shorten this to just:

```suggestion
	result := a + b
```

File: calculator.go [hunk 10]
```diff
@@ -1,13 +1,96 @@
...
+	for _, n := range nums {
+		product *= n
+	}
+	return product
+}
+
+// Divide takes a list of numbers and divides them.
+// Error is returned when dividing by 0.
+func Divide(a, b float64, nums ...float64) (float64, error) {
+	err := fmt.Errorf("cannot divide by zero")
```
Comment (bitfield) [inline calculator.go]: We can use `errors.New` if we just have a string constant.

File: calculator_test.go [hunk 11]
```diff
@@ -2,23 +2,170 @@ package calculator_test
...
+func TestDivide(t *testing.T) {
+	t.Parallel()
+	var testCases = []struct {
+		name        string
+		a, b        float64
+		nums        []float64
+		want        float64
+		errExpected bool
+	}{
+		{name: "positive", a: 2, b: 2, want: 1, errExpected: false},
```
Comment (bitfield) [inline calculator_test.go:129]: Try a test case like `2/3`, because that's a number that can't be accurately represented in floating-point.

File: calculator_test.go [hunk 12]
```diff
@@ -2,23 +2,170 @@ package calculator_test
...
+	"time"
 )
 
 func TestAdd(t *testing.T) {
 	t.Parallel()
-	var want float64 = 4
-	got := calculator.Add(2, 2)
-	if want != got {
-		t.Errorf("want %f, got %f", want, got)
+	var testCases = []struct {
```
Comment (bitfield) [inline calculator_test.go:13]: And we could test Add, Subtract, and Multiply all in the same test, couldn't we? Can you see how?

File: calculator_test.go [hunk 13]
```diff
@@ -2,23 +2,170 @@ package calculator_test
...
+		{name: "one zero", a: 0, b: 3, want: 0, errExpected: true},
+		{name: "two zeroes", a: 0, b: 0, want: 0, errExpected: true},
+		{name: "decimal", a: 2.5, b: 4, want: 0.625, errExpected: false},
+		{name: "many numbers", a: 1024, b: 2, nums: []float64{2, 2, 2, 2, 2, 2, 2, 2}, want: 2, errExpected: false},
+	}
+	for _, tc := range testCases {
+		got, err := calculator.Divide(tc.a, tc.b, tc.nums...)
+		errReceived := err != nil
+		if tc.errExpected != errReceived {
+			t.Fatalf("test name: %s unexpected error, got: %v", tc.name, err)
```
Comment (bitfield) [inline calculator_test.go]: "unexpected error _status_" is a good way to phrase this, since the unexpected event might be that we _didn't_ receive an error when we should have done!

**Review by bitfield — COMMENTED** [review]
Neat!

File: calculator_test.go (hunk 12 above)
Comment (maiku1008) [reply calculator_test.go:13]: Sure thing, we can pass a field that is a `func(a, b float64, nums ...float64) float64`, and call that one instead of each function separately.

I will push a separate `TestAllOperationsWithNoError` to demonstrate that.

Comment (maiku1008) [reply calculator_test.go:13]: Addressed in https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/2456ac63b78579d40850eaebb74ce1bf25b512d8

File: calculator_test.go (hunk 11 above)
Comment (maiku1008) [reply calculator_test.go:129]: Addressed in https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/48b6d6bff1d8adc4e283472874dfa3f90d002132

File: calculator.go [hunk 14]
```diff
@@ -1,13 +1,97 @@
...
+		result *= n
+	}
+	return result
+}
+
+// Divide takes a list of numbers and divides them.
+// Error is returned when dividing by 0.
+func Divide(a, b float64, nums ...float64) (float64, error) {
+	err := errors.New("cannot divide by zero")
+	if a == 0 || b == 0 {
```
Comment (bitfield) [inline calculator.go]: But if `a` is zero, that's not an error!

File: calculator.go [hunk 15]
```diff
@@ -1,13 +1,97 @@
...
+		return 0, err
+	}
+
+	var result float64 = a / b
+	for _, n := range nums {
+		if n == 0 || result == 0 {
```
Comment (bitfield) [inline calculator.go]: But if `result` is zero, that's not an error!

File: calculator.go [hunk 16]
```diff
@@ -1,13 +1,97 @@
...
+	var a, b float64
+	var operation string
+	var err error
+
+	operations := []string{"+", "-", "*", "/"}
+	for _, op := range operations {
+		if strings.Contains(input, op) {
+			operation = op
+			input = strings.Replace(input, op, " ", 1)
+			_, err = fmt.Sscanf(input, "%f%f", &a, &b)
```
Comment (bitfield) [inline calculator.go]: We could just call `fmt.Sscanf` directly, couldn't we, without using `strings.Contains` first.

**Review by bitfield — COMMENTED** [review]
Looking good so far!

File: calculator.go (hunk 14 above)
Comment (maiku1008) [reply calculator.go]: addressed in https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/7f0584f8ddefaae0f23bb0d586254ebb26619421

File: calculator.go (hunk 15 above)
Comment (maiku1008) [reply calculator.go]: :smirk: addressed in https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/7f0584f8ddefaae0f23bb0d586254ebb26619421

File: calculator.go (hunk 16 above)
Comment (maiku1008) [reply calculator.go]: I must admit this one got me stuck for a while, even the first time you suggested `fmt.Sscanf`.
I would love to do everything by calling it directly without any machinery.

The issue is that Sscanf will only work with space-separated values, which is the reason I went with the machinery that replaces whatever operator is found in the string (line 74), with a whitespace (line 76).

Without this, the tests for string values like `"2*2"` fail.
```
var a, b float64
var op string
operation := "2*2"
	
fmt.Sscanf(operation, "%f%s%f", &a, &op, &b)
fmt.Println(a, op, b) // 2 *2 0
```

A bit stumped here :)

Comment (bitfield) [reply calculator.go]: Well, the problem here is that Scanf is being too greedy, isn't it? You asked it to find a string starting at `*` and it did, but it went on to see how much more of a string it could read, and it read as far as `*2`. That leaves nothing for the second value.

Perhaps we could tell it to be less greedy? Is there a way to specify in the format string that we're only interested in a string of exactly one character for the operator? See what you can find out.

Comment (maiku1008) [reply calculator.go]: 😂 Thanks for that. Addressed in https://github.com/micuffaro/ftl-fundamentals/pull/1/commits/e8408efc091c65add548f0c73fdbc6201feac7a1

**Review by bitfield — APPROVED** [review]
Great job! You've earned yourself a yellow belt 😁
